### ISSUE TITLE
copilot: add reasoning using reinforcement learning through self-play (absolute-zero-reasoner method)

### ISSUE THREAD

**Issue by davidbasilefilho** [davidbasilefilho/atlas#8]
**Prompt Title:** Project: Implement Absolute-Zero-Reasoner for Atlas Model Optimization

**Role:** You are an expert AI programmer. Your task is to implement a reinforcement learning loop based on the "Absolute-Zero-Reasoner" paper to optimize the `Atlas` deep learning model.

**Context:**

  * **Target Project (`Atlas`):** A PyTorch implementation of a transformer-like model with a distinct memory architecture. Key components to understand from `/readme.md` are the `DeepMemoryModule`, `PolynomialFeatureMap`, and the `Omega Rule` for memory updates. The goal is to optimize the configuration of this specific architecture, not a standard Transformer.
  * **Core Methodology (`Absolute-Zero-Reasoner`):** A method for model improvement via self-play and Monte Carlo Tree Search (MCTS), detailed in `https://arxiv.org/pdf/2505.03335.pdf`. Your implementation must adapt its principles.
  * **Package Management:** The project uses `uv` from astral.sh for dependency management. Ensure all package installation and environment setup uses `uv` commands (e.g., `uv pip install -e .`).

**Primary Objective:**
Create a self-play optimization loop that intelligently searches for the optimal `AtlasConfig`. The "actions" in this self-play game are modifications to the model's configuration, guided by an MCTS process that uses structured reasoning to form hypotheses about the `Atlas` architecture.

-----

#### 1\. Create the Optimization Orchestrator (`optimize.py`)

This script will be the entry point for the MCTS self-play loop. It will manage the search tree, execute simulations, and select the best configurations.

#### 2\. Implement the Simulation and Reward Function

The simulation step involves training and evaluating an `Atlas` model with a given configuration. The results are then converted into a scalar reward to guide the MCTS search.

#### 3\. Data Collection for Reasoner Improvement

During self-play, log the data generated by the MCTS: the initial configuration (state), the MCTS search policy, and the final reward. This data is for potential future fine-tuning of the reasoner.

-----

#### MCTS Cycle Adaptation

This project adapts the paper's algorithm for configuration search. The MCTS cycle (Selection, Expansion, Simulation, Backpropagation) must be implemented as follows:

  * **State:** A node in the search tree represents a state, defined by a specific `AtlasConfig`.
  * **Action:** An action is a dictionary representing a mutation to the current node's `AtlasConfig` (e.g., `{"polynomial_degree": 4}`).
  * **Selection:** Traverse the tree from the root by repeatedly selecting the child node with the highest **PUCT** (Polynomial Upper Confidence Trees) score until a leaf node is reached. The PUCT formula is:
    $PUCT(s, a) = Q(s, a) + P(s, a) \\cdot c\_{puct} \\cdot \\frac{\\sqrt{\\sum\_b N(s, b)}}{1 + N(s, a)}$
      - $Q(s, a)$ is the average reward of the action.
      - $P(s, a)$ is the prior probability of selecting the action (can be uniform for the first implementation).
      - $N(s, a)$ is the visit count for the action.
      - $c\_{puct}$ is a constant controlling exploration (e.g., 1.5).
  * **Expansion & Reasoning:** When a leaf node is reached, act as the **reasoner**. Generate $N$ new candidate configurations. Your reasoning in the `<R>...</R>` tags **must be specific to the Atlas architecture**. Formulate a clear hypothesis connecting the parameter change to a specific model component.
    > **Example Reasoner Output:**
    > ` <R>The current  `context\_window\_size`is small, limiting the`Omega Rule` 's ability to perform long-range memory updates. I will test increasing it to 1024 to improve performance on the needle-in-haystack task. As a counterpoint, I will also test increasing  `memory\_depth`to see if a deeper`DeepMemoryModule`  can compensate for a smaller context window.</R> `
    > `[`
    > `   {"context_window_size": 1024}, `
    > `   {"memory_depth": 3} `
    > `]`
  * **Simulation (Empirical Rollout):** This is a key adaptation from the paper. Since we lack a pre-trained value network, the "simulation" to determine a node's value is an **empirical test**. You must execute a fast training and evaluation run of the `Atlas` model using the node's configuration. The scalar reward from this test serves as the simulation result `v`.
  * **Backpropagation:** After a simulation, update the statistics ($N(s,a)$ and $Q(s,a)$) for all nodes along the path from the simulated leaf back to the root.

#### `optimize.py` Structure and Communication

  - Implement the MCTS logic and a `Node` class as previously specified.
  - Use Python's `subprocess` module to call `main.py` for each simulation to ensure process isolation.
  - Modify `main.py` to accept a path to a configuration JSON file and to print a final JSON object with evaluation metrics to `stdout`.

#### Reward Calculation

Implement a clear reward function to convert evaluation metrics into a scalar value.

```python
def calculate_reward(metrics: dict) -> float:
    """Calculates a scalar reward from evaluation metrics."""
    # Normalize perplexity (lower is better).
    ppl = metrics.get("perplexity", 100.0)
    ppl_score = max(0, 1 - (ppl / 100.0))
    # Needle accuracy is already 0-1.
    needle_acc = metrics.get("needle_accuracy", 0.0)
    # Normalize memory capacity score.
    mem_score = metrics.get("memory_capacity_score", 0.0) / 100.0
    # Weighted sum for the final reward.
    reward = (0.4 * ppl_score) + (0.4 * needle_acc) + (0.2 * mem_score)
    return reward
```

-----

### Deliverables

1.  A new Python script: **`optimize.py`** containing the MCTS optimization loop.
2.  Modifications to existing project files (**`main.py`**, **`evaluate.py`**) to allow for programmatic execution and to output a final JSON summary.
3.  A `README.md` section explaining how to run `optimize.py` and detailing the MCTS adaptation for this project.
4.  A file **`reasoner_training_data.jsonl`** containing at least 10 examples of the collected `(state, policy, reward)` data.
